### PR TITLE
API: enhance the service grouping

### LINKED ISSUE #271: API: make public API endpoints easy to setup
# 🚀 Feature Request

In order to make public API endpoints easy to set up, we must do everything to:
- minimize the work required to setup the environment (just use a reverse proxy with a certificate and point it to the host/port/path and that's it)
- to maximize the compatibility between providers

Now there are two public servers providing Semux API:
- https://api.semux.online (+ https://api.testnet.semux.online) 
- https://www.sempy.online/api/ui/#/semux (this is almost 100% Semux API)

Both use very different approach and software:
- https://github.com/witoldsz/api.semux.online
- https://github.com/mdodong/semuxapi-py

These projects were created because Semux Core is missing dedicated endpoint serving nothing but public API, where one:
- does not need to authenticate
- does not have access to wallet of the Semux node
- does not see non-public endpoints in the swagger-generated page and the specification file

## Current solution

In my opinion, the _API: divide APIs into groups_ #267 is not a good solution:
- does little to help setup a public API endpoint,
- does assume that from now on, each and every public endpoint must be GET and wallet endpoint must NOT be GET,
- is confusing, e.g. the extra comment in config file: 
    ```
    # Allow GET methods (safe for public access) only
    api.allowGetMethodsOnly = true
    ```
    for no reason in the Semux API jargon, `allowGetMethodsOnly` equals `publicMethodsOnly`

## Suggestions

1. rollback the GET=public rule
2. make it so that once we setup reverse proxy, there is nothing but public API with no wallet endpoints, no authentication sections, etc…

We can either:
- setup separate public API from wallet API (two swagger files, two root paths)
- or just mark each endpoint (public or wallet) and let the Semux Core generate the proper API page+endpints according to some settings like `api.publicOnly=true`.

The first approach has an advantage that one can publish public API (e.g. reverse proxy points to host:port/api/public) and still use wallet API internally.

The latter approach is implemented in https://github.com/witoldsz/api.semux.online by filtering the swagger specification file. @orogvany said:

> Swagger has API so no manual filtering necessary

### PR DESCRIPTION
**Pull request by semuxgo**
## Description

This PR introduces the concept of public and private services, to partially resolve #271.

## Test Plan

Existing unit tests

## Related issues and/or PRs

#271

### CODE REVIEW

File: src/main/java/org/semux/api/ApiHandlerImpl.java [hunk 1]
```diff
@@ -85,6 +113,25 @@ public Response service(HttpMethod method, String path, Map<String, String> para
         }
     }
 
+    @Override
+    public boolean isAuthRequired(HttpMethod method, String path) {
+        Route route = matchRoute(method, path);
+        return route == null ? false : !route.isPublic;
```
Comment (witoldsz) [inline src/main/java/org/semux/api/ApiHandlerImpl.java]: `return route != null && !route.isPublic` is it better?

File: src/main/java/org/semux/api/ApiHandlerImpl.java [hunk 2]
```diff
@@ -85,6 +113,25 @@ public Response service(HttpMethod method, String path, Map<String, String> para
...
+    }
+
+    /**
+     * Matches route by [version, method, path]
+     */
+    private Route matchRoute(HttpMethod method, String path) {
+        Route route = null;
+        Matcher m = VERSIONED_PATH.matcher(path);
+        if (m.matches() && routes.containsKey(ApiVersion.of(m.group(1)))) {
```
Comment (witoldsz) [inline src/main/java/org/semux/api/ApiHandlerImpl.java]: named groups would be less confusing than some `1` and `2`…

**Comment by witoldsz** [comment]
So, the concept here does not cover the case where I would like to use my node as a public Semux API node and still be able to access wallet API, is that right? If so, that means the public Semux nodes are, in practice, not accessible for owners (no GUI, no API).

**Comment by semuxgo** [comment]
I think you can do both now. The private APIs (wallet) are enabled and protected by basic auth, while the public APIs are just open to public, without basic auth.

**Comment by witoldsz** [comment]
OK, but I would feel uncomfortable exposing everything and having just the auth as the final shield from accessing my wallet. Of course I could set up some extended matching rules, not to allow certain endpoints to go through (black/white listed in reverse proxy service), but it complicates the deployment.

Last, but not least, when exposing public API I would like the swagger definition to be trimmed down to only what's available. See the https://api.semux.online. You won't find wallet endpoints there or the authentication schemes definition. And I can still see and use full version when bypassing the proxy.

**Comment by honeycrypto** [comment]
+1 for the dynamic swagger page, but what will be the best implementation for this? Desktop users want to see all available endpoints, even the protected ones (because they are on localhost). While public API providers want to show only public endpoints in swagger UI. 
Another config option `api.showPrivateInSwaggerUI = true`? 

2) I think we should encourage users do not ever use wallets with public API as their personal wallets. If there is a bug or whatever, they can lose their funds. If the wallet has always 0 balance, it's not that critical if someone once will get a chance to access private API. 

3) We can provide a documentation how to setup a public api node from scratch, including some kind of `nginx.conf` with the blacklisted private endpoints, instruction on getting let's encrypt certificates and other requited steps. 

TL;DR: i'm ok with exposing to internet password protected endpoints. Dynamic Swagger UI is good to have to minimise required configs to proxy/firewall and certificates.

**Comment by semuxgo** [comment]
Dynamic swagger generation would require extra engineering work. Will consider this in another PR.

Displaying all APIs (even though some of them are inaccessible) won't necessarily be a bad thing. I personally use https://api.semux.online not only for calling these endpoints but also for reading the API specification.
